rfcs/QueryLevelNullability.md

```diff
@@ -0,0 +1,348 @@
+For example if you are trying to render an information page for a movie, you won't be able to do that if the name field
+of the movie is missing. In that case it would be preferable to fail as early as possible.
+
+According to the official GraphQL best practice, this field should be non-nullable:
+
+> When designing a GraphQL schema, it's important to keep in mind all the problems that could go wrong and if "null" is
+> an appropriate value for a failed field. Typically it is, but occasionally, it's not. In those cases, use non-null
+> types to make that guarantee.
+
+However, the same field may not be "vital" for every feature in the application that fetches it. Marking the field as
+non-null in the schema would result in those other features erroring unnecessarily whenever the field is "null".
+
+#### When handling partial results
+
+It is often unclear how to handle partial results.
+
+- What elements are essential to having an adequate developer experience? How many elements can fail before you 
+  should just show an error message instead?
+- When any combination of elements can fail, it is very hard to anticipate all edge cases and how the UI 
+  might look and work, including transitions to other screens.
+
+Implementing these decisions significantly complicates the client-side logic for handling query results.
+
+### ✨ Examples
+
+#### Non-nullable field
+```graphql
+query GetBusinessName($id: String!) {
+  business(id: $id) {
+    name!
+  }
+}
+```
+would codegen to the following type on iOS.
+```swift
+struct GetBusinessNameQuery {
+  let id: String
+
+  struct Data {
+    let business: Business?
+
+    struct Business {
+      /// Lack of `?` indicates that `name` will never be `null`
+      let name: String
+    }
+  }
+}
+```
+
+#### Non-nullable object with nullable fields
+```graphql
+query GetBusinessReviews {
+  business(id: 4) {
+    reviews! {
+      rating
+      text
+    }
+  }
+}
+```
+
+## ✅ RFC Goals
+
+- Non-nullable syntax that is based off of syntax that developers will already be familiar with
+- Enable GraphQL codegen tools to generate more ergonomic types
+
+## 🚫 RFC Non-goals
+
+## 🗳️ Alternatives considered
+
+### A `@nonNull` official directive
+
+This solution offers the same benefits as the proposed solution. Since many GraphQL codegen tools already support the `@skip` and `@include` directives, this solution likely has a faster turnaround.
+
+### A `@nonNull` custom directive
+
+This is an alternative being used at some of the companies represented in this proposal for the time being.
+
+While this solution simplifies some client-side logic, it does not meaningfully improve the developer experience for clients.
+
+* The cache implementations of GraphQL client libraries also need to understand the custom directive to behave correctly. Currently, when a client library caches a null field based on an operation without a directive, it will return the null field for another operation with this directive.
```

Comment:
This will be true even with custom syntax. For composability of fragments, it would be worrisome if this new syntax made it impossible to have these two fragments in the same application:
```
fragment UserOkWithoutName on User {
  name
}
fragment UserMustHaveName on User {
  name!
}
```
With consistency, I could end up re-rendering a component built with `UserMustHaveName` based on a response fetched using only `UserOkWithoutName`.

And in fact it gets worse:
```
fragment BestFriendWithoutName on User {
  best_friend {
    id
  }
}
fragment BestFriendWithName on User {
  best_friend {
    id
    name!
  }
}
```
If my best friend, when I fetched `BestFriendWithName`, is `User:1`, and then suddenly, after fetching `BestFriendWithoutName` is `User:2`, I no longer am capable of rendering the component relying on `BestFriendWithName` (because `name` is *unset* on `User:1`). If the client library doesn't know what to do in that case, you can end up with consistency updates causing crashes.

Reply:
As you said, the client library will need to understand this new syntax/directive. I am not familiar with GraphQL client implementations, so I am not sure how difficult my imagined behavior will be to implement. Doesn't this behavior solve the issue you outlined?

> If a GraphQL client finds that a component relies on `BestFriendWithName `, but the cached data doesn't have `name`, it should not return any of the cached data. To the client, it is like a state before fetching started. Depending on `fetch-policy` (to use Apollo's terminology), the client can start a new request to fetch the missing field data.

Reply:
> Doesn't this behavior solve the issue you outlined?

That's how *Apollo* solves it, but that strategy isn't true universally. For instance that's not how Relay solves it. The client I work on *can't* solve it this way, as our consistency subscriptions are at the response level, for queries that include hundreds of fragments (whether that's a good choice or not is irrelevant: it's how things work for the largest GraphQL client at Facebook): if missing `name` meant the entire consistency update should not be served, whenever a product chose to mark their field with `!`, they could cause entire unrelated surfaces to lose consistency.

Another valid strategy would also be to return the data, but with `best_friend` being null: all the other data in the fragment/consistency subscription would still be delivered (this is Relay's strategy for `@required(action: NONE)`). Or requiring the product to null check every piece of data, always, regardless of its nullability annotation.

Reply:
Do you mean that Relay requires field selections on the same id and type to behave the same globally? If so, I can see how my suggested behavior won't work. If not:

> Another valid strategy would also be to return the data, but with best_friend being null: all the other data in the fragment/consistency subscription would still be delivered (this is Relay's strategy for @required(action: NONE))

This is what I imagined. The client would nullify up until the most immediate nullable parent.

Reply:
@mjmahone thank you for the feedback, this is great for us to think about! ❤️  I think I see the problem you've laid out here, but let me know if I've misunderstood anything!

when a not-required field is fetched and returned as `null`, we could end up wiping away values in the cache that are required by other features.

for features using codegen types, the Swift/Kotlin/etc. properties corresponding to those "required" fields will be non-Optional. the value will be replaced by `null` in the cache -- which would result in a crash anytime those non-Optional properties are accessed.

is it feasible for the client to distinguish in the response between these two cases (e.g. perhaps via the `errors` array)?
* fields that are `null` due to an error, and
* fields that are `null` because they are also `null` in the database

in the former case, the client library could choose to not clear the existing value in the cache. from some experimentation, this doesn't seem to be how e.g. the Apollo iOS cache works with regular nullable fields -- if a `null` value is returned, it seems to replace any value that was previously in the cache (I wasn't able to verify that this was also the case when an error occurs on that field).

is there any other way for the client library to know what to do in this case?